File: imagingbook-common/src/main/java/imagingbook/common/math/eigen/accord/QZHES.java
package imagingbook.common.math.eigen.accord;

public abstract class QZHES {
	
	// EISPACK Routines, see http://www.netlib.org/eispack/

	/// <summary>
	///   Adaptation of the original Fortran QZHES routine from EISPACK.
	/// </summary>
	/// <remarks>
	///   This subroutine is the first step of the qz algorithm
	///   for solving generalized matrix eigenvalue problems,
	///   Siam J. Numer. anal. 10, 241-256(1973) by Moler and Stewart.
	///
	///   This subroutine accepts a pair of real general matrices and
	///   reduces one of them to upper Hessenberg form and the other
	///   to upper triangular form using orthogonal transformations.
	///   it is usually followed by  qzit,  qzval  and, possibly,  qzvec.
	///   
	///   For the full documentation, please check the original function.
	/// </remarks>
	static int qzhes(int n, double[][] a, double[][] b, boolean matz, double[][] z) {
		
//		System.out.println("runnin qzhes");
		int i, j, k, l, nm1, nm2;
		double r, s, t;
		int l1;
		double u1, u2, v1, v2;
		int lb, nk1;
		double rho;
	
		if (matz) {
			// If we are interested in computing the
			// eigenvectors, set Z to identity(n,n)
			for (j = 0; j < n; j++) {
				for (i = 0; i < n; i++)
					z[i][j] = 0.0;
				z[j][j] = 1.0;
			}
		}
	
		// reduce b to upper triangular form
		if (n <= 1) {
			return 0;
		}
		
		nm1 = n - 1;
		for (l = 0; l < nm1; l++) {
			l1 = l + 1;
			s = 0.0;
	
			for (i = l1; i < n; i++)
				s = s + (Math.abs(b[i][l]));
	
			if (s == 0.0) {
				continue;
			}
			s = s + Math.abs(b[l][l]);
			r = 0.0;
	
			for (i = l; i < n; ++i) {
				// Computing 2nd power
				b[i][l] = b[i][l] / s;
				r = r + b[i][l] * b[i][l];
			}
	
			r = Special.Sign(Math.sqrt(r), b[l][l]);
			b[l][l] = b[l][l] + r;
			rho = r * b[l][l];
	
			for (j = l1; j < n; j++) {
				t = 0.0;
				for (i = l; i < n; i++)
					t += b[i][l] * b[i][j];
				t = -t / rho;
				for (i = l; i < n; i++) {
					b[i][j] = b[i][j] + t * b[i][l];
				}
			}
	
			for (j = 0; j < n; j++) {
				t = 0.0;
				for (i = l; i < n; i++) {
					t += b[i][l] * a[i][j];
				}
				t = -t / rho;
				for (i = l; i < n; i++) {
					a[i][j] = a[i][j] + t * b[i][l];
				}
			}
	
			b[l][l] = -s * r;
			for (i = l1; i < n; i++) {
				b[i][l] = 0.0;
			}
		}
	
		// reduce a to upper Hessenberg form, while keeping b triangular
		if (n == 2) {
			return 0;
		}
		nm2 = n - 2;
		for (k = 0; k < nm2; k++) {
			nk1 = nm1 - k - 1; // = n - 2 - k;
	
			// for l=n-1 step -1 until k+1 do
			for (lb = 0; lb < nk1; lb++) {
				l = n - lb - 2;
				l1 = l + 1;
	
				// zero a(l+1,k)
				s = (Math.abs(a[l][k])) + (Math.abs(a[l1][k]));
	
				if (s == 0.0) {
					continue;
				}
				
				u1 = a[l][k] / s;
				u2 = a[l1][k] / s;
//				r = Special.Sign(Math.sqrt(u1 * u1 + u2 * u2), u1); // Special.Sign
				r = Math.copySign(Math.hypot(u1, u2), u1); 
				v1 = -(u1 + r) / r;
				v2 = -u2 / r;
				u2 = v2 / v1;
	
				for (j = k; j < n; j++) {
					t = a[l][j] + u2 * a[l1][j];
					a[l][j] = a[l][j] + t * v1;
					a[l1][j] = a[l1][j] + t * v2;
				}
	
				a[l1][k] = 0.0;
	
				for (j = l; j < n; j++) {
					t = b[l][j] + u2 * b[l1][j];
					b[l][j] = b[l][j] + t * v1;
					b[l1][j] = b[l1][j] + t * v2;
				}
	
				// Zero b(l+1,l)
				s = (Math.abs(b[l1][l1])) + (Math.abs(b[l1][l]));
				if (s == 0.0) {
					continue;
				}
				u1 = b[l1][l1] / s;
				u2 = b[l1][l] / s;
//				r = Special.Sign(Math.sqrt(u1 * u1 + u2 * u2), u1);
				r = Math.copySign(Math.hypot(u1, u2), u1);
				v1 = -(u1 + r) / r;
				v2 = -u2 / r;
				u2 = v2 / v1;
	
				for (i = 0; i <= l1; i++) {
					t = b[i][l1] + u2 * b[i][l];
					b[i][l1] = b[i][l1] + t * v1;
					b[i][l] = b[i][l] + t * v2;
				}
	
				b[l1][l] = 0.0;
	
				for (i = 0; i < n; i++) {
					t = a[i][l1] + u2 * a[i][l];
					a[i][l1] = a[i][l1] + t * v1;
					a[i][l] = a[i][l] + t * v2;
				}
	
				if (matz) {
					for (i = 0; i < n; i++) {
						t = z[i][l1] + u2 * z[i][l];
						z[i][l1] = z[i][l1] + t * v1;
						z[i][l] = z[i][l] + t * v2;
					}
				}
			}
		}
	
//		System.out.println("done qzhes");
		return 0;
	} // end of qzhes

}

File: imagingbook-common/src/main/java/imagingbook/common/math/eigen/accord/eispack-routines/qzhes.f90

subroutine qzhes ( n, a, b, matz, z )

!*****************************************************************************80
!
!! QZHES carries out transformations for a generalized eigenvalue problem.
!
!  Discussion:
!
!    This subroutine is the first step of the QZ algorithm
!    for solving generalized matrix eigenvalue problems.
!
!    This subroutine accepts a pair of real general matrices and
!    reduces one of them to upper Hessenberg form and the other
!    to upper triangular form using orthogonal transformations.
!    it is usually followed by QZIT, QZVAL and, possibly, QZVEC.
!
!  Licensing:
!
!    This code is distributed under the GNU LGPL license.
!
!  Modified:
!
!    18 October 2009
!
!  Author:
!
!    Original FORTRAN77 version by Smith, Boyle, Dongarra, Garbow, Ikebe,
!    Klema, Moler.
!    FORTRAN90 version by John Burkardt.
!
!  Reference:
!
!    James Wilkinson, Christian Reinsch,
!    Handbook for Automatic Computation,
!    Volume II, Linear Algebra, Part 2,
!    Springer, 1971,
!    ISBN: 0387054146,
!    LC: QA251.W67.
!
!    Brian Smith, James Boyle, Jack Dongarra, Burton Garbow,
!    Yasuhiko Ikebe, Virginia Klema, Cleve Moler,
!    Matrix Eigensystem Routines, EISPACK Guide,
!    Lecture Notes in Computer Science, Volume 6,
!    Springer Verlag, 1976,
!    ISBN13: 978-3540075462,
!    LC: QA193.M37.
!
!  Parameters:
!
!    Input, integer ( kind = 4 ) N, the order of the matrices.
!
!    Input/output, real ( kind = 8 ) A(N,N).  On input, the first real general
!    matrix.  On output, A has been reduced to upper Hessenberg form.  The
!    elements below the first subdiagonal have been set to zero.
!
!    Input/output, real ( kind = 8 ) B(N,N).  On input, a real general matrix.
!    On output, B has been reduced to upper triangular form.  The elements
!    below the main diagonal have been set to zero.
!
!    Input, logical MATZ, should be TRUE if the right hand transformations
!    are to be accumulated for later use in computing eigenvectors.
!
!    Output, real ( kind = 8 ) Z(N,N), contains the product of the right hand
!    transformations if MATZ is TRUE.
!
  implicit none

  integer ( kind = 4 ) n

  real ( kind = 8 ) a(n,n)
  real ( kind = 8 ) b(n,n)
  integer ( kind = 4 ) i
  integer ( kind = 4 ) j
  integer ( kind = 4 ) k
  integer ( kind = 4 ) l
  integer ( kind = 4 ) l1
  integer ( kind = 4 ) lb
  logical matz
  integer ( kind = 4 ) nk1
  integer ( kind = 4 ) nm1
  real ( kind = 8 ) r
  real ( kind = 8 ) rho
  real ( kind = 8 ) s
  real ( kind = 8 ) t
  real ( kind = 8 ) u1
  real ( kind = 8 ) u2
  real ( kind = 8 ) v1
  real ( kind = 8 ) v2
  real ( kind = 8 ) z(n,n)
!
!  Set Z to the identity matrix.
!
  if ( matz ) then

    z(1:n,1:n) = 0.0D+00

    do i = 1, n
      z(i,i) = 1.0D+00
    end do

  end if
!
!  Reduce B to upper triangular form.
!
  if ( n <= 1 ) then
    return
  end if

  nm1 = n - 1

  do l = 1, n - 1

    l1 = l + 1

    s = sum ( abs ( b(l+1:n,l) ) )

    if ( s /= 0.0D+00 ) then

      s = s + abs ( b(l,l) )
      b(l:n,l) = b(l:n,l) / s

      r = sqrt ( sum ( b(l:n,l)**2 ) )
      r = sign ( r, b(l,l) )
      b(l,l) = b(l,l) + r
      rho = r * b(l,l)

      do j = l + 1, n

        t = dot_product ( b(l:n,l), b(l:n,j) )

        b(l:n,j) = b(l:n,j) - t * b(l:n,l) / rho

      end do

      do j = 1, n

        t = dot_product ( b(l:n,l), a(l:n,j) )

        a(l:n,j) = a(l:n,j) - t * b(l:n,l) / rho

      end do

      b(l,l) = - s * r
      b(l+1:n,l) = 0.0D+00

    end if

  end do
!
!  Reduce A to upper Hessenberg form, while keeping B triangular.
!
  if ( n == 2 ) then
    return
  end if

  do k = 1, n - 2

     nk1 = nm1 - k

     do lb = 1, nk1

        l = n - lb
        l1 = l + 1
!
!  Zero A(l+1,k).
!
        s = abs ( a(l,k) ) + abs ( a(l1,k) )

        if ( s /= 0.0D+00 ) then

        u1 = a(l,k) / s
        u2 = a(l1,k) / s
        r = sign ( sqrt ( u1**2 + u2**2 ), u1 )
        v1 = - ( u1 + r) / r
        v2 = - u2 / r
        u2 = v2 / v1

        do j = k, n
          t = a(l,j) + u2 * a(l1,j)
          a(l,j) = a(l,j) + t * v1
          a(l1,j) = a(l1,j) + t * v2
        end do

        a(l1,k) = 0.0D+00

        do j = l, n
          t = b(l,j) + u2 * b(l1,j)
          b(l,j) = b(l,j) + t * v1
          b(l1,j) = b(l1,j) + t * v2
        end do
!
!  Zero B(l+1,l).
!
        s = abs ( b(l1,l1) ) + abs ( b(l1,l) )

        if ( s /= 0.0 ) then

          u1 = b(l1,l1) / s
          u2 = b(l1,l) / s
          r = sign ( sqrt ( u1**2 + u2**2 ), u1 )
          v1 =  -( u1 + r ) / r
          v2 = -u2 / r
          u2 = v2 / v1

          do i = 1, l1
            t = b(i,l1) + u2 * b(i,l)
            b(i,l1) = b(i,l1) + t * v1
            b(i,l) = b(i,l) + t * v2
          end do

          b(l1,l) = 0.0D+00

          do i = 1, n
            t = a(i,l1) + u2 * a(i,l)
            a(i,l1) = a(i,l1) + t * v1
            a(i,l) = a(i,l) + t * v2
          end do

          if ( matz ) then

            do i = 1, n
              t = z(i,l1) + u2 * z(i,l)
              z(i,l1) = z(i,l1) + t * v1
              z(i,l) = z(i,l) + t * v2
            end do

          end if

        end if

      end if

    end do

  end do

  return
end